pkg/platform/api/graphql/model/buildplanner/buildExpression.go

```diff
@@ -0,0 +1,267 @@
+
+	var requirements []Requirement
+	err = json.Unmarshal(requirementsData, &requirements)
+	if err != nil {
+		return nil, errs.Wrap(err, "Could not unmarshal JSON")
+	}
+
+	return requirements, nil
+}
+
+func (bx BuildExpression) Update(operation Operation, requirement Requirement) error {
+	var err error
+	switch operation {
+	case OperationAdd:
+		err = bx.AddRequirement(requirement)
+	case OperationRemove:
+		err = bx.RemoveRequirement(requirement)
+	case OperationUpdate:
+		err = bx.UpdateRequirement(requirement)
+	default:
+		return errs.New("Unsupported operation")
+	}
+	if err != nil {
+		return errs.Wrap(err, "Could not update BuildExpression's requirements")
+	}
+
+	err = bx.UpdateTimestamp()
+	if err != nil {
+		return errs.Wrap(err, "Could not update BuildExpression's timestamp")
+	}
+
+	return nil
+}
+
+func (bx BuildExpression) AddRequirement(requirement Requirement) error {
+	solveNode, err := getSolveNode(bx)
+	if err != nil {
+		return errs.Wrap(err, "Could not get solve node")
+	}
+
+	requirementsNode, err := getRequirementsNode(bx)
+	if err != nil {
+		return errs.Wrap(err, "Could not get requirements node")
+	}
+
+	newRequirementData, err := json.Marshal(requirement)
+	if err != nil {
+		return errs.Wrap(err, "Could not marshal JSON")
+	}
+
+	var newRequirement Requirement
+	err = json.Unmarshal(newRequirementData, &newRequirement)
+	if err != nil {
+		return errs.Wrap(err, "Could not unmarshal JSON")
+	}
+
+	requirementsNode = append(requirementsNode, newRequirement)
+
+	solveNode[RequirementsKey] = requirementsNode
+
+	return nil
+}
+
+func (bx BuildExpression) RemoveRequirement(requirement Requirement) error {
+	solveNode, err := getSolveNode(bx)
+	if err != nil {
+		return errs.Wrap(err, "Could not get solve node")
+	}
+
+	requirementsNode, err := getRequirementsNode(bx)
+	if err != nil {
+		return errs.Wrap(err, "Could not get requirements node")
+	}
+
+	for i, req := range requirementsNode {
+		r, ok := req.(map[string]interface{})
+		if !ok {
+			return errs.New("Requirement in BuildExpression is malformed")
+		}
+
+		if r[RequirementNameKey] == requirement.Name && r[RequirementNamespaceKey] == requirement.Namespace {
+			requirementsNode = append(requirementsNode[:i], requirementsNode[i+1:]...)
+			solveNode[RequirementsKey] = requirementsNode
+			return nil
+		}
+	}
+
+	return errs.New("Could not find requirement")
+}
+
+func (bx BuildExpression) UpdateRequirement(requirement Requirement) error {
+	solveNode, err := getSolveNode(bx)
+	if err != nil {
+		return errs.Wrap(err, "Could not get solve legacy node")
+	}
+
+	requirementsNode, err := getRequirementsNode(bx)
+	if err != nil {
+		return errs.Wrap(err, "Could not get requirements node")
+	}
+
+	for i, req := range requirementsNode {
+		r, ok := req.(map[string]interface{})
+		if !ok {
+			return errs.New("Requirement in BuildExpression is malformed")
+		}
+
+		if r[RequirementNameKey] == requirement.Name && r[RequirementNamespaceKey] == requirement.Namespace {
```

Comment:
I'm not following... we have to iterate through the list of requirements to find the one we want to update.

Reply:
Sorry. Line 186 is raising an error. Ideally this would be done earlier like in `NewBuildExpression()`. Also verify that `r[RequirementNameKey]` exists rather than panics.